btree/btree.c

```diff
@@ -4335,112 +4869,254 @@ static void vkvv_cut(const struct nd *node, int idx, int size)
+	struct dir_rec   *rec   = vkvv_get_dir_addr(slot->s_node);
+	m0_bcount_t       hsize = sizeof(*h) - sizeof(h->vkvv_opaque);
+
+	void *key_addr;
+	int   ksize;
+	int  *p_ksize = &ksize;
+	void *val_addr;
+	int   vsize;
+	int  *p_vsize = &vsize;
+	void *dir_addr;
+	int   dsize;
+	int  *p_dsize = &dsize;
+
+	if (slot->s_idx < h->vkvv_used) {
+		key_addr = vkvv_key(slot->s_node, slot->s_idx);
+		val_addr = vkvv_val(slot->s_node, h->vkvv_used);
+		if (node_level(slot->s_node) != 0)
+			val_addr -= INT_OFFSET;
+
+		dir_addr = rec + slot->s_idx;
+		vkvv_calc_size_for_capture(slot, h->vkvv_used, p_ksize, p_vsize,
+					   p_dsize);
+		M0_BTREE_TX_CAPTURE(tx, seg, key_addr, ksize);
+		M0_BTREE_TX_CAPTURE(tx, seg, val_addr, vsize);
+		if (node_level(slot->s_node) == 0)
+			M0_BTREE_TX_CAPTURE(tx, seg, dir_addr, dsize);
+
+	} else if (h->vkvv_opaque == NULL)
+		hsize += sizeof(h->vkvv_opaque);
+
+	M0_BTREE_TX_CAPTURE(tx, seg, h, hsize);
 }
 
-static bool vkvv_verify(const struct nd *node)
+static void vkvv_node_alloc_credit(const struct nd *node,
+				struct m0_be_tx_credit *accum)
 {
-	/**
-	 * @brief This function verify the data in the node.
-	 */
+	int             node_size   = node->n_size;
+	int             shift       = __builtin_ffsl(node_size) - 1;
+
+	m0_be_allocator_credit(NULL, M0_BAO_ALLOC_ALIGNED,
+			       node_size, shift, accum);
 }
 
-static void vkvv_opaque_set(const struct segaddr *addr, void *opaque)
+static void vkvv_node_free_credit(const struct nd *node,
+				struct m0_be_tx_credit *accum)
 {
-	/**
-	 * @brief This function saves the opaque data.
-	 */
+	int             node_size   = node->n_size;
+	int             shift       = __builtin_ffsl(node_size) - 1;
+	int             header_size = sizeof(struct vkvv_head);
+
+	m0_be_allocator_credit(NULL, M0_BAO_FREE_ALIGNED,
+			       node_size, shift, accum);
+
+	m0_be_tx_credit_add(accum, &M0_BE_TX_CREDIT(1, header_size));
 }
 
-static void *vkvv_opaque_get(const struct segaddr *addr)
+static void vkvv_rec_put_credit(const struct nd *node, m0_bcount_t ksize,
+			      m0_bcount_t vsize,
+			      struct m0_be_tx_credit *accum)
 {
-	/**
-	 * @brief This function return the opaque data.
-	 */
+	int             node_size   = node->n_size;
+
+	m0_be_tx_credit_add(accum, &M0_BE_TX_CREDIT(3, node_size));
 }
 
-static void vkvv_capture(struct slot *slot, struct m0_be_tx *tx)
+static void vkvv_rec_update_credit(const struct nd *node, m0_bcount_t ksize,
+				 m0_bcount_t vsize,
+				 struct m0_be_tx_credit *accum)
 {
-	/**
-	 * @brief This function will capture the data in node segment.
-	 */
+	int             node_size   = node->n_size;
+
+	m0_be_tx_credit_add(accum, &M0_BE_TX_CREDIT(3, node_size));
+}
+
+static void vkvv_rec_del_credit(const struct nd *node, m0_bcount_t ksize,
+			      m0_bcount_t vsize,
+			      struct m0_be_tx_credit *accum)
+{
+	int             node_size   = node->n_size;
+
+	m0_be_tx_credit_add(accum, &M0_BE_TX_CREDIT(3, node_size));
```

Comment:
instead of 3 it should be 4 as you are calling CAPTURE at max 4 times in vkvv_capture. same is applicable for update, put credit calculations

Reply:
Updated